### Carga inicial de datos

In [ ]:
import pandas as pd
import os

tables = {}
files = os.listdir('../data/raw')
print(files)
for file in files:
    if file.endswith('.csv'):
        table_name = file.replace('.csv', '')
        table_path = os.path.join('../data/raw', file)
        tables[table_name] = pd.read_csv(table_path)
        print(f"Loaded {table_name} with shape {tables[table_name].shape}")

aisles = tables['aisles']
departments = tables['departments']
orders = tables['orders']
products = tables['products']
order_products_prior = tables['order_products__prior']
order_products_train = tables['order_products__train']

['.gitkeep', 'aisles.csv', 'departments.csv', 'orders.csv', 'order_products__prior.csv', 'order_products__train.csv', 'products.csv']
Loaded aisles with shape (134, 2)
Loaded departments with shape (21, 2)
Loaded orders with shape (3421083, 7)
Loaded order_products__prior with shape (32434489, 4)
Loaded order_products__train with shape (1384617, 4)
Loaded products with shape (49688, 4)


In [ ]:
aisles = tables['aisles'].copy()
print("Primeras filas:" )
print(aisles.head())

print("Ultimas filas: ")
print(aisles.tail())

print("Información del DataFrame:")
print(aisles.info())

print("Cantidad de filas y columnas:")
print(aisles.shape)

print("Valores nulos por columna:")
print(aisles.isnull().sum())

print("Porcentaje de valores nulos por columna:")
print(aisles.isnull().mean() * 100)

print("Duplicados por columna:")
print(aisles.duplicated().sum())

print("Claves unicas por culumna:")
print(aisles.nunique())

print("Frecuencia de claves unicas por columna:")
print(aisles.value_counts())

print("Estadísticas descriptivas:")
print(aisles.describe())

print("Tipos de datos del DataFrame:")
print(aisles.dtypes)

Primeras filas:
   aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation
Ultimas filas: 
     aisle_id                       aisle
129       130    hot cereal pancake mixes
130       131                   dry pasta
131       132                      beauty
132       133  muscles joints pain relief
133       134  specialty wines champagnes
Información del DataFrame:
<class 'pandas.DataFrame'>
RangeIndex: 134 entries, 0 to 133
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   aisle_id  134 non-null    int64
 1   aisle     134 non-null    str  
dtypes: int64(1), str(1)
memory usage: 4.2 KB
None
Cantidad de filas y columnas:
(134, 2)
Valores nulos por columna:
aisle_id    0
aisle       0
dtype: int64
Porcentaje de valores nulos por columna:
aisle_id 

## Auditoría de calidad — `aisles`

La tabla `aisles` contiene el catálogo de pasillos o categorías intermedias utilizadas para clasificar los productos del dataset de Instacart.

### Estructura
- Filas: **134**
- Columnas: **2**
- `aisle_id`: identificador numérico del pasillo.
- `aisle`: nombre descriptivo del pasillo.

### Valores nulos
No se encontraron valores nulos en ninguna de las columnas:

- `aisle_id`: 0 nulos
- `aisle`: 0 nulos

Por lo tanto, no es necesario realizar imputaciones ni eliminar registros por valores faltantes.

### Duplicados
No se encontraron filas duplicadas.

Además:

- `aisle_id` posee **134 valores únicos**.
- `aisle` posee **134 valores únicos**.

Esto indica que existe una correspondencia uno a uno entre cada identificador y cada nombre de pasillo.

### Validación de la clave
`aisle_id` funciona correctamente como clave primaria de la tabla.

Los identificadores se encuentran en el rango:

- Mínimo: **1**
- Máximo: **134**

No se detectaron identificadores repetidos ni valores fuera del rango observado.

### Tipos de datos
Los tipos de datos son adecuados:

- `aisle_id`: `int64`
- `aisle`: `str`

No se requieren conversiones de tipos.

### Conclusión
La tabla `aisles` presenta una buena calidad de datos:

- sin valores nulos;
- sin registros duplicados;
- clave primaria única;
- nombres de pasillos únicos;
- tipos de datos adecuados.

**No se requiere ninguna transformación o limpieza sobre esta tabla en esta etapa.**

In [ ]:
departments = tables['departments'].copy()
print("Primeras filas:" )
print(departments.head())

print("Ultimas filas: ")
print(departments.tail())

print("Información del DataFrame:")
print(departments.info())

print("Cantidad de filas y columnas:")
print(departments.shape)

print("Valores nulos por columna:")
print(departments.isnull().sum())

print("Porcentaje de valores nulos por columna:")
print(departments.isnull().mean() * 100)

print("Duplicados por columna:")
print(departments.duplicated().sum())

print("Claves unicas por culumna:")
print(departments.nunique())

print("Frecuencia de claves unicas por columna:")
print(departments.value_counts())

print("Estadísticas descriptivas:")
print(departments.describe())

print("Tipos de datos del DataFrame:")
print(departments.dtypes)

Primeras filas:
   department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol
Ultimas filas: 
    department_id department
16             17  household
17             18     babies
18             19     snacks
19             20       deli
20             21    missing
Información del DataFrame:
<class 'pandas.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   department_id  21 non-null     int64
 1   department     21 non-null     str  
dtypes: int64(1), str(1)
memory usage: 638.0 bytes
None
Cantidad de filas y columnas:
(21, 2)
Valores nulos por columna:
department_id    0
department       0
dtype: int64
Porcentaje de valores nulos por columna:
department_id    0.0
department       0.0
dtype: float64
Duplicados por columna:
0
Claves unicas por culumna:
department_id    

## Auditoría de calidad — `departments`

La tabla `departments` contiene el catálogo de departamentos utilizados para agrupar los productos del dataset de Instacart en categorías de nivel superior.

### Estructura
- Filas: **21**
- Columnas: **2**
- `department_id`: identificador numérico del departamento.
- `department`: nombre descriptivo del departamento.

### Valores nulos
No se encontraron valores nulos en ninguna de las columnas:

- `department_id`: 0 nulos
- `department`: 0 nulos

No es necesario realizar imputaciones ni eliminar registros.

### Duplicados
No se encontraron filas duplicadas.

Además:

- `department_id` posee **21 valores únicos**.
- `department` posee **21 valores únicos**.

Esto indica una correspondencia uno a uno entre cada identificador y cada nombre de departamento.

### Validación de la clave
`department_id` funciona correctamente como clave primaria de la tabla.

Los identificadores se encuentran en el rango:

- Mínimo: **1**
- Máximo: **21**

No se detectaron identificadores repetidos.

### Tipos de datos
Los tipos de datos son adecuados:

- `department_id`: `int64`
- `department`: `str`

No se requieren conversiones de tipos.

### Observación sobre la categoría `missing`
Existe un departamento denominado `missing` con `department_id = 21`.

Este valor no representa un dato nulo a nivel técnico, sino una categoría explícita presente en el dataset. Por lo tanto, no debe tratarse como `NaN` ni eliminarse durante la limpieza.

Más adelante será conveniente revisar qué productos están asociados a este departamento para determinar su impacto en el análisis.

### Conclusión
La tabla `departments` presenta buena calidad de datos:

- sin valores nulos;
- sin filas duplicadas;
- clave primaria única;
- nombres de departamentos únicos;
- tipos de datos adecuados.

**No se requiere limpieza sobre esta tabla en esta etapa.**


In [ ]:
order_products_prior = tables['order_products__prior'].copy()
print("Primeras filas:" )
print(order_products_prior.head())

print("Ultimas filas: ")
print(order_products_prior.tail())

print("Información del DataFrame:")
print(order_products_prior.info())

print("Cantidad de filas y columnas:")
print(order_products_prior.shape)

print("Valores nulos por columna:")
print(order_products_prior.isnull().sum())

print("Porcentaje de valores nulos por columna:")
print(order_products_prior.isnull().mean() * 100)

print("Duplicados por columna:")
print(order_products_prior.duplicated().sum())

print("Claves unicas por culumna:")
print(order_products_prior.nunique())

print("Frecuencia de claves unicas por columna:")
print(order_products_prior.value_counts())

print("Estadísticas descriptivas:")
print(order_products_prior.describe())

print("Tipos de datos del DataFrame:")
print(order_products_prior.dtypes)

print("Integridad referencial entre order_products_prior y orders:")
print(order_products_prior['order_id'].isin(orders['order_id']).all())

print("Integridad referencial entre order_products_prior y products:")
print(order_products_prior['product_id'].isin(products['product_id']).all())

print("Duplicados en la combinación de order_id y product_id:")
print( order_products_prior.duplicated(
    subset=['order_id', 'product_id']
).sum())

print("Duplicados en la combinación de order_id y add_to_cart_order:")
print(
    order_products_prior.duplicated(
    subset=['order_id', 'add_to_cart_order']
).sum()
)


Primeras filas:
   order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0
Ultimas filas: 
          order_id  product_id  add_to_cart_order  reordered
32434484   3421083       39678                  6          1
32434485   3421083       11352                  7          0
32434486   3421083        4600                  8          0
32434487   3421083       24852                  9          1
32434488   3421083        5020                 10          1
Información del DataFrame:
<class 'pandas.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 4 columns):
 #   Column             Dtype
---  ------             -----
 0   order_id           int64
 1   product_id         int64
 2   add_to_cart_order  int6

## Auditoría de calidad — `order_products__prior`

La tabla `order_products__prior` contiene los productos incluidos en los pedidos históricos de los usuarios de Instacart. Cada fila representa la aparición de un producto dentro de un pedido.

Esta tabla constituye una de las principales tablas transaccionales del proyecto y será fundamental para analizar recompra y asociaciones entre productos.

### Estructura
- Filas: **32.434.489**
- Columnas: **4**
- Pedidos únicos: **3.214.874**
- Productos únicos observados: **49.677**

Las columnas son:

- `order_id`: identificador del pedido.
- `product_id`: identificador del producto.
- `add_to_cart_order`: posición en la que el producto fue agregado al carrito.
- `reordered`: indicador binario que señala si el usuario ya había comprado anteriormente ese producto.

### Valores nulos
No se encontraron valores nulos en ninguna columna:

- `order_id`: 0
- `product_id`: 0
- `add_to_cart_order`: 0
- `reordered`: 0

Por lo tanto, no es necesario realizar imputaciones ni eliminar registros por datos faltantes.

### Duplicados
No se encontraron filas completamente duplicadas.

También se verificó la unicidad de las combinaciones:

- `order_id + product_id`
- `order_id + add_to_cart_order`

No se encontraron duplicados en estas combinaciones, por lo que un mismo producto no aparece repetido dentro de un pedido y cada posición del carrito corresponde a un único producto.

### Variable `reordered`
La variable `reordered` contiene únicamente los valores:

- `0`: producto no marcado como recompra.
- `1`: producto marcado como recompra.

No se encontraron valores fuera del dominio esperado.

La media observada es aproximadamente **0,5897**, lo que indica que alrededor del **58,97 % de las líneas de producto de los pedidos históricos corresponden a productos marcados como recompra**.

Este resultado será relevante posteriormente para el análisis de comportamiento de recompra.

### Posición en el carrito
`add_to_cart_order` presenta valores desde:

- Mínimo: **1**
- Máximo: **145**

La posición comienza correctamente en 1 y no se detectaron valores iguales o menores a 0.

El máximo de 145 indica la existencia de pedidos de gran tamaño, pero por sí solo no constituye una anomalía.

### Integridad referencial
Se verificaron las relaciones con las tablas principales.

Todos los `order_id` presentes en `order_products__prior` existen en:

`orders.order_id`

Asimismo, todos los `product_id` presentes en la tabla existen en:

`products.product_id`

Por lo tanto, no se detectaron registros huérfanos.

### Productos del catálogo
En los pedidos históricos aparecen **49.677 productos únicos**, mientras que el catálogo `products` contiene **49.688 productos**.

Esto significa que existen **11 productos del catálogo que no aparecen en `order_products__prior`**.

Esta diferencia no representa necesariamente un problema de calidad, ya que esos productos podrían aparecer en otros subconjuntos del dataset, como `order_products__train`.

### Tipos de datos
Todas las columnas fueron cargadas como `int64`:

- `order_id`: `int64`
- `product_id`: `int64`
- `add_to_cart_order`: `int64`
- `reordered`: `int64`

Los tipos son válidos para el análisis.

### Conclusión
La tabla `order_products__prior` presenta una alta calidad de datos:

- sin valores nulos;
- sin filas duplicadas;
- sin duplicación lógica de productos dentro de pedidos;
- posiciones de carrito consistentes;
- variable `reordered` dentro del dominio esperado;
- integridad referencial completa con `orders` y `products`.

**No se requiere limpieza correctiva de la tabla en esta etapa.**

In [ ]:
order_products_train = tables['order_products__train'].copy()
print("Primeras filas:" )
print(order_products_train.head())

print("Ultimas filas: ")
print(order_products_train.tail())

print("Información del DataFrame:")
print(order_products_train.info())

print("Cantidad de filas y columnas:")
print(order_products_train.shape)

print("Valores nulos por columna:")
print(order_products_train.isnull().sum())

print("Porcentaje de valores nulos por columna:")
print(order_products_train.isnull().mean() * 100)

print("Duplicados por columna:")
print(order_products_train.duplicated().sum())

print("Claves unicas por culumna:")
print(order_products_train.nunique())

print("Frecuencia de claves unicas por columna:")
print(order_products_train.value_counts())

print("Estadísticas descriptivas:")
print(order_products_train.describe())

print("Tipos de datos del DataFrame:")
print(order_products_train.dtypes)

print("Integridad referencial entre order_products_train y orders:")
print(order_products_train['order_id'].isin(orders['order_id']).all())

print("Integridad referencial entre order_products_train y products:")
print(order_products_train['product_id'].isin(products['product_id']).all())

train_order_ids = orders.loc[
    orders['eval_set'] == 'train',
    'order_id'
]

print("Integridad referencial entre order_products_train y train_order_ids:")
print(order_products_train['order_id'].isin(
    train_order_ids
).all())

print("Número de órdenes de entrenamiento únicas:")
print(orders[orders['eval_set'] == 'train']['order_id'].nunique())

print("Duplicados en la combinación de order_id y add_to_cart_order:")
print(order_products_train.duplicated(
    subset=['order_id', 'add_to_cart_order']
).sum())

print("Duplicados en la combinación de order_id y product_id:")
print(order_products_train.duplicated(
    subset=['order_id', 'product_id']
).sum())

Primeras filas:
   order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1
Ultimas filas: 
         order_id  product_id  add_to_cart_order  reordered
1384612   3421063       14233                  3          1
1384613   3421063       35548                  4          1
1384614   3421070       35951                  1          1
1384615   3421070       16953                  2          1
1384616   3421070        4724                  3          1
Información del DataFrame:
<class 'pandas.DataFrame'>
RangeIndex: 1384617 entries, 0 to 1384616
Data columns (total 4 columns):
 #   Column             Non-Null Count    Dtype
---  ------             --------------    -----
 0   order_id           1384617 non-null  int64
 1   product

## Auditoría de calidad — `order_products__train`

La tabla `order_products__train` contiene los productos incluidos en los pedidos pertenecientes al conjunto de entrenamiento de Instacart. Cada fila representa un producto dentro de un pedido.

### Estructura
- Filas: **1.384.617**
- Columnas: **4**
- Órdenes únicas: **131.209**
- Productos únicos observados: **39.123**

Las columnas son:

- `order_id`: identificador del pedido.
- `product_id`: identificador del producto.
- `add_to_cart_order`: posición en la que el producto fue agregado al carrito.
- `reordered`: indicador de si el producto había sido comprado anteriormente por el usuario.

### Valores nulos
No se encontraron valores nulos en ninguna columna.

Por lo tanto, no se requieren imputaciones ni eliminación de registros por datos faltantes.

### Duplicados
No se encontraron filas completamente duplicadas.

También se verificó la unicidad de las combinaciones:

- `order_id + product_id`
- `order_id + add_to_cart_order`

No se detectaron duplicaciones lógicas dentro de los pedidos.

### Variable `reordered`
La variable contiene únicamente los valores esperados:

- `0`: producto no marcado como recompra.
- `1`: producto marcado como recompra.

La media de `reordered` es aproximadamente **0,5986**, por lo que cerca del **59,86 % de las líneas de producto corresponden a productos marcados como recompra**.

### Posición en el carrito
`add_to_cart_order` presenta valores entre:

- Mínimo: **1**
- Máximo: **80**

No se detectaron posiciones iguales o inferiores a 0.

### Integridad referencial
Se verificó que:

- Todos los `order_id` existen en `orders`.
- Todos los `product_id` existen en `products`.
- Todos los pedidos de esta tabla corresponden a registros con `eval_set = 'train'` en `orders`.

Por lo tanto, no existen registros huérfanos ni pedidos asignados incorrectamente al conjunto de entrenamiento.

### Tipos de datos
Todas las columnas utilizan `int64`, un tipo adecuado para los identificadores y variables numéricas presentes.

### Conclusión
La tabla `order_products__train` presenta una buena calidad de datos:

- sin valores nulos;
- sin filas duplicadas;
- sin duplicaciones lógicas dentro de los pedidos;
- valores válidos en `reordered`;
- posiciones de carrito consistentes;
- integridad referencial completa con `orders` y `products`.

**No se requiere limpieza correctiva sobre esta tabla en esta etapa.**

In [ ]:
orders = tables['orders'].copy()
print("Primeras filas:" )
print(orders.head())

print("Ultimas filas: ")
print(orders.tail())

print("Información del DataFrame:")
print(orders.info())

print("Cantidad de filas y columnas:")
print(orders.shape)

print("Valores nulos por columna:")
print(orders.isnull().sum())

print("Porcentaje de valores nulos por columna:")
print(orders.isnull().mean() * 100)

print("Duplicados por columna:")
print(orders.duplicated().sum())

print("Claves unicas por culumna:")
print(orders.nunique())

print("Frecuencia de claves unicas por columna:")
print(orders.value_counts())

print("Estadísticas descriptivas:")
print(orders.describe())

print("Tipos de datos del DataFrame:")
print(orders.dtypes)

# order_id debe ser clave primaria
print("order_id es clave primaria:")
print(orders['order_id'].is_unique)

# Dominio de eval_set
print("Dominio de eval_set:")
print(orders['eval_set'].value_counts())

# Rangos válidos
print("Rangos válidos de order_dow:")
print(orders['order_dow'].between(0, 6).all())
print("Rangos válidos de order_hour_of_day:")
print(orders['order_hour_of_day'].between(0, 23).all())

print("Rangos válidos de days_since_prior_order:")
print(orders['days_since_prior_order'].dropna().between(0, 30).all())

# Los nulos deberían corresponder al primer pedido

print("Nulos en el primer pedido:")
print(orders.loc[
    orders['order_number'] == 1,
    'days_since_prior_order'
].isna().all())

# Y los pedidos posteriores no deberían tener NaN

print("Nulos en los pedidos posteriores:")
print(orders.loc[
    orders['order_number'] > 1,
    'days_since_prior_order'
].notna().all())

Primeras filas:
   order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
Ultimas filas: 
         order_id  user_id eval_set  order_number  order_dow  \
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231

## Auditoría de calidad — `orders`

La tabla `orders` contiene el historial de pedidos de los usuarios de Instacart y permite reconstruir la secuencia de compras de cada usuario.

### Estructura
- Filas: **3.421.083**
- Columnas: **7**
- Usuarios únicos: **206.209**
- Pedidos únicos: **3.421.083**

Las columnas son:

- `order_id`: identificador único del pedido.
- `user_id`: identificador del usuario.
- `eval_set`: subconjunto al que pertenece el pedido (`prior`, `train` o `test`).
- `order_number`: posición del pedido dentro del historial del usuario.
- `order_dow`: día de la semana codificado entre 0 y 6.
- `order_hour_of_day`: hora del día codificada entre 0 y 23.
- `days_since_prior_order`: cantidad de días transcurridos desde el pedido anterior.

### Valores nulos
La única columna con valores nulos es:

- `days_since_prior_order`: **206.209 nulos**
- Porcentaje aproximado: **6,03 %**

Se comprobó que:

- todos los pedidos con `order_number = 1` tienen `days_since_prior_order = NaN`;
- todos los pedidos con `order_number > 1` tienen un valor registrado en `days_since_prior_order`.

Por lo tanto, estos valores nulos son **estructurales y esperados**: el primer pedido de cada usuario no tiene un pedido anterior contra el cual calcular la cantidad de días transcurridos.

No deben imputarse ni eliminarse.

### Duplicados y claves
No se encontraron filas completamente duplicadas.

`order_id` contiene **3.421.083 valores únicos**, coincidiendo con la cantidad total de registros, por lo que funciona correctamente como clave primaria de la tabla.

`user_id` contiene **206.209 usuarios únicos** y puede repetirse, ya que cada usuario puede realizar múltiples pedidos.

### Validación de rangos
Se verificaron los dominios de las variables temporales.

#### `order_dow`
Los valores se encuentran dentro del rango esperado:

`0 – 6`

No se detectaron valores fuera de rango.

#### `order_hour_of_day`
Los valores se encuentran dentro del rango:

`0 – 23`

No se detectaron horas inválidas.

#### `days_since_prior_order`
Los valores no nulos se encuentran dentro del rango:

`0 – 30 días`

No se detectaron valores fuera del dominio esperado.

### Tipos de datos
Los tipos detectados son:

- `order_id`: `int64`
- `user_id`: `int64`
- `eval_set`: `str`
- `order_number`: `int64`
- `order_dow`: `int64`
- `order_hour_of_day`: `int64`
- `days_since_prior_order`: `float64`

Los tipos son adecuados para el análisis.

`order_dow` y `order_hour_of_day` se mantienen como enteros porque representan códigos numéricos de día y hora, no fechas completas.

`days_since_prior_order` se mantiene como `float64` debido a la presencia de valores `NaN` en los primeros pedidos.

### Conclusión
La tabla `orders` presenta una alta calidad de datos:

- sin duplicados;
- clave primaria válida;
- rangos temporales correctos;
- tipos de datos adecuados;
- valores nulos explicados por la lógica del negocio.

**No se requiere limpieza correctiva en esta etapa.**

In [ ]:
products = tables['products'].copy()

print("Primeras filas:")
print(products.head())

print("Ultimas filas:")
print(products.tail())

print("Información del DataFrame:")
print(products.info())

print("Cantidad de filas y columnas:")
print(products.shape)

print("Valores nulos por columna:")
print(products.isnull().sum())

print("Porcentaje de valores nulos por columna:")
print(products.isnull().mean() * 100)

print("Duplicados por columna:")
print(products.duplicated().sum())

print("Claves unicas por columna:")
print(products.nunique())

print("Frecuencia de claves unicas por columna:")
print(products.value_counts())

print("Estadísticas descriptivas:")
print(products.describe())

print("Tipos de datos del DataFrame:")
print(products.dtypes)


# product_id debe ser clave primaria

print("product_id es clave primaria:")
print(products['product_id'].is_unique)


# Verificar integridad referencial con aisles

print("Integridad referencial entre products y aisles:")
print(
    products['aisle_id'].isin(
        aisles['aisle_id']
    ).all()
)


# Verificar integridad referencial con departments

print("Integridad referencial entre products y departments:")
print(
    products['department_id'].isin(
        departments['department_id']
    ).all()
)


# Verificar IDs válidos

print("Todos los product_id son mayores a 0:")
print((products['product_id'] > 0).all())

print("Todos los aisle_id son mayores a 0:")
print((products['aisle_id'] > 0).all())

print("Todos los department_id son mayores a 0:")
print((products['department_id'] > 0).all())


# Revisar nombres de productos repetidos

print("Cantidad de nombres de productos duplicados:")
print(
    products.duplicated(
        subset=['product_name']
    ).sum()
)


# Mostrar nombres repetidos en caso de que existan

print("Productos con nombres repetidos:")
print(
    products[
        products.duplicated(
            subset=['product_name'],
            keep=False
        )
    ].sort_values('product_name')
)


# Revisar productos pertenecientes al departamento 'missing'

missing_department_ids = departments.loc[
    departments['department'].str.lower() == 'missing',
    'department_id'
]

print("Cantidad de productos pertenecientes al departamento 'missing':")
print(
    products[
        products['department_id'].isin(missing_department_ids)
    ].shape[0]
)

print("Ejemplos de productos pertenecientes al departamento 'missing':")
print(
    products[
        products['department_id'].isin(missing_department_ids)
    ].head(10)
)

missing_products = products[
    products['department_id'] == 21
]

print("Aisles utilizados por productos del departamento missing:")
print(missing_products['aisle_id'].value_counts())
aisles[aisles['aisle_id'] == 100]

Primeras filas:
   product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  
0             19  
1             13  
2              7  
3              1  
4             13  
Ultimas filas:
       product_id                                product_name  aisle_id  \
49683       49684   Vodka, Triple Distilled, Twist of Vanilla       124   
49684       49685          En Croute Roast Hazelnut Cranberry        42   
49685       49686                            Artisan Baguette       112   
49686       49687  Smartblend Healthy Metabolism Dry Cat Food        41   
49687      

,aisle_id,aisle
99,100,missing


## Auditoría de calidad — `products`

La tabla `products` contiene el catálogo de productos de Instacart y relaciona cada producto con un pasillo (`aisle`) y un departamento.

### Estructura
- Filas: **49.688**
- Columnas: **4**
- Productos únicos: **49.688**
- Pasillos utilizados: **134**
- Departamentos utilizados: **21**

Las columnas son:

- `product_id`: identificador único del producto.
- `product_name`: nombre del producto.
- `aisle_id`: identificador del pasillo al que pertenece.
- `department_id`: identificador del departamento al que pertenece.

### Valores nulos
No se encontraron valores nulos en ninguna columna:

- `product_id`: 0
- `product_name`: 0
- `aisle_id`: 0
- `department_id`: 0

Por lo tanto, no es necesario realizar imputaciones.

### Duplicados
No se encontraron filas completamente duplicadas.

Además:

- `product_id` contiene **49.688 valores únicos**.
- `product_name` contiene **49.688 valores únicos**.

No existen nombres de productos duplicados en el catálogo.

### Clave primaria
`product_id` funciona correctamente como clave primaria.

Todos sus valores son únicos y mayores a 0.

### Integridad referencial

Se verificó que:

- todos los `aisle_id` presentes en `products` existen en `aisles`;
- todos los `department_id` presentes en `products` existen en `departments`.

Por lo tanto, no existen productos con referencias huérfanas hacia las tablas de clasificación.

### Validación de identificadores
Todos los identificadores presentan valores positivos:

- `product_id > 0`
- `aisle_id > 0`
- `department_id > 0`

Además, los rangos observados son consistentes con las tablas relacionadas:

- `product_id`: **1 – 49.688**
- `aisle_id`: **1 – 134**
- `department_id`: **1 – 21**

### Productos sin clasificación

Se identificaron **1.258 productos** asociados al:

- `department_id = 21` → `missing`
- `aisle_id = 100` → `missing`

Esto confirma que el dataset utiliza una categoría explícita denominada `missing` para representar productos cuya clasificación en pasillo y departamento no está disponible.

Estos registros:

- no contienen valores nulos reales;
- mantienen integridad referencial;
- poseen `product_id` y `product_name` válidos;
- forman parte legítima del catálogo.

Por lo tanto, **no deben eliminarse ni imputarse**.

Se conservarán como una categoría propia y se tendrá en cuenta esta limitación en los análisis posteriores por pasillo o departamento.

### Tipos de datos

Los tipos encontrados son adecuados:

- `product_id`: `int64`
- `product_name`: `str`
- `aisle_id`: `int64`
- `department_id`: `int64`

No se requieren conversiones.

### Conclusión

La tabla `products` presenta una buena calidad de datos:

- sin valores nulos;
- sin filas duplicadas;
- clave primaria válida;
- nombres de productos únicos;
- integridad referencial completa con `aisles` y `departments`;
- identificadores dentro de rangos válidos.

**No se requiere limpieza correctiva.**


In [ ]:
dim_products = (
    products
    .merge(
        aisles,
        on='aisle_id',
        how='left',
        validate='many_to_one'
    )
    .merge(
        departments,
        on='department_id',
        how='left',
        validate='many_to_one'
    )
)

print(dim_products.head())
print(dim_products.shape)
print(dim_products.isnull().sum())

print("Misma cantidad de productos:")
print(len(dim_products) == len(products))

print("Sin categorías faltantes después del join:")
print(
    dim_products[['aisle', 'department']]
    .notna()
    .all()
    .all()
)

order_basket_summary = (
    order_products_prior
    .groupby('order_id')
    .agg(
        basket_size=('product_id', 'size'),
        unique_products=('product_id', 'nunique'),
        reordered_items=('reordered', 'sum')
    )
    .reset_index()
)

order_basket_summary['reorder_rate'] = (
    order_basket_summary['reordered_items']
    / order_basket_summary['basket_size']
)

prior_orders = orders[
    orders['eval_set'] == 'prior'
][[
    'order_id',
    'user_id',
    'order_number',
    'order_dow',
    'order_hour_of_day',
    'days_since_prior_order'
]]

order_basket_summary = order_basket_summary.merge(
    prior_orders,
    on='order_id',
    how='left',
    validate='one_to_one'
)

print(order_basket_summary.head())
print(order_basket_summary.shape)

print("Pedidos sin user_id:")
print(order_basket_summary['user_id'].isnull().sum())

product_reorder_metrics = (
    order_products_prior
    .groupby('product_id')
    .agg(
        total_purchases=('order_id', 'size'),
        total_orders=('order_id', 'nunique'),
        reordered_purchases=('reordered', 'sum'),
        avg_cart_position=('add_to_cart_order', 'mean')
    )
    .reset_index()
)

product_reorder_metrics['reorder_rate'] = (
    product_reorder_metrics['reordered_purchases']
    / product_reorder_metrics['total_purchases']
)

product_reorder_metrics = product_reorder_metrics.merge(
    dim_products,
    on='product_id',
    how='left',
    validate='one_to_one'
)

product_reorder_metrics = product_reorder_metrics[[
    'product_id',
    'product_name',
    'aisle_id',
    'aisle',
    'department_id',
    'department',
    'total_purchases',
    'total_orders',
    'reordered_purchases',
    'reorder_rate',
    'avg_cart_position'
]]

aisle_reorder_metrics = (
    product_reorder_metrics
    .groupby(
        ['aisle_id', 'aisle'],
        as_index=False
    )
    .agg(
        total_purchases=('total_purchases', 'sum'),
        reordered_purchases=('reordered_purchases', 'sum'),
        products=('product_id', 'nunique')
    )
)

aisle_reorder_metrics['reorder_rate'] = (
    aisle_reorder_metrics['reordered_purchases']
    / aisle_reorder_metrics['total_purchases']
)

department_reorder_metrics = (
    product_reorder_metrics
    .groupby(
        ['department_id', 'department'],
        as_index=False
    )
    .agg(
        total_purchases=('total_purchases', 'sum'),
        reordered_purchases=('reordered_purchases', 'sum'),
        products=('product_id', 'nunique')
    )
)

department_reorder_metrics['reorder_rate'] = (
    department_reorder_metrics['reordered_purchases']
    / department_reorder_metrics['total_purchases']
)

import os

processed_path = '../data/processed'

os.makedirs(
    processed_path,
    exist_ok=True
)

dim_products.to_csv(
    '../data/processed/dim_products.csv',
    index=False
)

order_basket_summary.to_csv(
    '../data/processed/order_basket_summary.csv',
    index=False
)

product_reorder_metrics.to_csv(
    '../data/processed/product_reorder_metrics.csv',
    index=False
)

aisle_reorder_metrics.to_csv(
    '../data/processed/aisle_reorder_metrics.csv',
    index=False
)

department_reorder_metrics.to_csv(
    '../data/processed/department_reorder_metrics.csv',
    index=False
)

print("Archivos processed generados correctamente.")
print(os.listdir('../data/processed'))

   product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id                       aisle department  
0             19               cookies cakes     snacks  
1             13           spices seasonings     pantry  
2              7                         tea  beverages  
3              1                frozen meals     frozen  
4             13  marinades meat preparation     pantry  
(49688, 6)
product_id       0
product_name     0
aisle_id         0
department_id    0
aisle            0
department       0
dtype: int64
Misma cantidad de productos:
True
Sin categor